subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c

```diff
@@ -1003,15 +881,65 @@ gst_msdkdec_finish_task (GstMsdkDec * thiz, MsdkDecTask * task)
     }
 
     if (G_LIKELY (frame)) {
-      if (G_LIKELY (surface->copy.buffer == NULL)) {
+      if (!thiz->do_copy) {
         /* gst_video_decoder_finish_frame will call gst_buffer_make_writable
-         * we need this to avoid copy buffer                               */
+         * we need this to avoid copy buffer                              */
         GST_MINI_OBJECT_FLAG_SET (surface->buf, GST_MINI_OBJECT_FLAG_LOCKABLE);
         frame->output_buffer = gst_buffer_ref (surface->buf);
       } else {
-        gst_video_frame_copy (&surface->copy, &surface->data);
-        frame->output_buffer = gst_buffer_ref (surface->copy.buffer);
-        unmap_frame (thiz, surface);
+        /* We need to do the copy from video memory to system memory */
+        GstBuffer *buffer = NULL;
+        GstVideoFrame src_frame;
+        GstVideoFrame dst_frame;
+        GstVideoInfo *src_info;
+        GstVideoInfo dst_info;
+        GstVideoCodecState *output_state =
+            gst_video_decoder_get_output_state (GST_VIDEO_DECODER (thiz));
+
+        src_info = &output_state->info;
+        gst_video_info_set_format (&dst_info, GST_VIDEO_INFO_FORMAT (src_info),
+            GST_VIDEO_INFO_WIDTH (src_info), GST_VIDEO_INFO_HEIGHT (src_info));
+
+        if (!gst_buffer_pool_is_active (thiz->other_pool) &&
+            !gst_buffer_pool_set_active (thiz->other_pool, TRUE)) {
+          GST_ERROR_OBJECT (thiz, "Failed to activate buffer pool");
+          return GST_FLOW_ERROR;
+        }
+
+        if (gst_buffer_pool_acquire_buffer (thiz->other_pool, &buffer, NULL)
+            != GST_FLOW_OK) {
+          GST_ERROR ("Failed to acquire buffer from pool");
+          gst_video_codec_state_unref (output_state);
+          return GST_FLOW_ERROR;
+        }
+
+        if (!gst_video_frame_map (&src_frame, src_info, surface->buf,
+                GST_MAP_READ)) {
+          GST_ERROR_OBJECT (thiz, "Failed to map buf to src frame");
+          gst_buffer_unref (buffer);
+          gst_video_codec_state_unref (output_state);
+          return GST_FLOW_ERROR;
+        }
+
+        if (!gst_video_frame_map (&dst_frame, &dst_info, buffer, GST_MAP_WRITE)) {
+          GST_ERROR_OBJECT (thiz, "Failed to map buf to dst frame");
+          gst_buffer_unref (buffer);
+          gst_video_codec_state_unref (output_state);
+          gst_video_frame_unmap (&src_frame);
+          return GST_FLOW_ERROR;
+        }
+
+        if (!gst_video_frame_copy (&dst_frame, &src_frame)) {
+          GST_ERROR_OBJECT (thiz, "Failed to copy surface data");
+          gst_video_frame_unmap (&src_frame);
+          gst_video_frame_unmap (&dst_frame);
+          return GST_FLOW_ERROR;
+        }
+        frame->output_buffer = gst_buffer_ref (buffer);
+        gst_video_frame_unmap (&src_frame);
+        gst_video_frame_unmap (&dst_frame);
+        gst_buffer_unref (buffer);
+        gst_video_codec_state_unref (output_state);
```

Comment:
It is better to write a static function like xxx_copy to handle this. And make the code a little clean.

Reply:
Integrate these multiple lines as a help function named _copy_to_sys_mem. Thanks for the advice.